feat: Enhanced performance monitoring and telemetry infrastructure

eLyiN opened the pull request:
## Summary

This PR implements comprehensive performance monitoring capabilities for the Gemini CLI as proposed in issue #2127. The implementation extends the existing OpenTelemetry infrastructure with detailed startup instrumentation, memory usage tracking, and advanced metrics collection.

## What's Changed

### 🚀 Detailed Startup Performance Monitoring
- **Phase-by-phase timing** for all CLI startup phases
- **Rich attributes** including auth type, telemetry settings, workspace info
- **Comprehensive instrumentation** covering settings loading, cleanup, extensions, config loading, service initialization
- Helps diagnose startup issues like #1685

### 🧠 Comprehensive Memory Monitoring
- **Real-time memory tracking** with new `MemoryMonitor` class
- **Automatic snapshots** at key lifecycle points (startup, major operations)
- **Component-specific monitoring** and memory growth analysis
- **Detailed heap statistics** including RSS, external memory, heap usage
- Helps detect memory issues like #1687

### 📊 Enhanced Metrics Framework
- **12 new performance metrics** following OpenTelemetry patterns
- **Performance scoring** and regression detection capabilities
- **Memory usage, CPU usage, tool execution breakdown** metrics
- **Startup timing, token efficiency, API request breakdown** tracking

## Technical Implementation

- ✅ **Extends existing OpenTelemetry** infrastructure without breaking changes
- ✅ **Follows established patterns** (`gemini_cli.*` naming convention)
- ✅ **Integrates with current configuration** (local/GCP targets)
- ✅ **Performance overhead designed** to be <1% of total execution
- ✅ **Maintains API compatibility** and existing telemetry controls

## Files Changed

- `packages/core/src/telemetry/constants.ts`: 12 new performance metrics constants
- `packages/core/src/telemetry/metrics.ts`: Enhanced recording functions (344 lines)
- `packages/core/src/telemetry/memory-monitor.ts`: New memory monitoring infrastructure (263 lines)
- `packages/core/src/telemetry/index.ts`: Updated exports for new functionality
- `packages/cli/src/gemini.tsx`: Startup performance instrumentation

## Testing

- ✅ **TypeScript compilation**: All packages compile without errors
- ✅ **Build verification**: Both CLI and Core packages build successfully
- ✅ **API compatibility**: No breaking changes to existing telemetry API

## Benefits

1. **Better issue diagnosis**: Resolve startup issues like #1685 with detailed timing data
2. **Memory issue detection**: Catch memory problems like #1687 before they impact users
3. **Performance optimization**: Provide data for optimizing CLI performance at scale
4. **Backwards compatible**: Seamless integration with existing telemetry infrastructure

## Usage

The performance monitoring is automatically enabled when telemetry is enabled:

```bash
# Enable telemetry to get performance monitoring
gemini --telemetry

# Or via environment
export GEMINI_TELEMETRY_ENABLED=true
```

Performance metrics are collected alongside existing telemetry data and sent to the same OTLP endpoints.

## Related Issues

- Closes #2127 
- Helps diagnose #1685 (startup/exit issues)
- Helps detect #1687 (memory issues)

## Notes

This implementation provides the foundation for performance monitoring. Future enhancements could include real-time dashboards and automated alerting as discussed in #2127.

umairidris commented:
/gemini review

eLyiN commented:
/gemini review

eLyiN commented:
/gemini review

eLyiN commented:
/gemini review

jerop reviewed (changes requested):
@eLyiN thank you for making these improvements to OpenTelemetry

Please update the docs and add tests

eLyiN commented:
Hi @jerop, thank you for reviewing my pull request. I've made the requested changes and also fixed some TypeScript issues that were causing problems with the `npm run preflight` command. Ready for review.

jerop commented:
@eLyiN apologies for the delay here, could you please rebase this change and will prioritize getting it in?

jerop reviewed (changes requested):
Getting this error when I try this out:

```
INT value type cannot accept a floating-point value for gemini_cli.startup.duration, ignoring the  fractional digits.
```

It looks like the recordStartupPerformance function in packages/core/src/telemetry/metrics.ts is receiving a floating-point number for the duration, but the underlying metric startupTimeHistogram is configured to only accept integers.

jacob314 commented on packages/core/src/telemetry/memory-monitor.ts:
```diff
@@ -0,0 +1,317 @@
+  /**
+   * Start continuous memory monitoring
+   */
+  start(config: Config, intervalMs: number = 5000): void {
```
Thank you for adding this! I am really excited to get analytics to understand how bad user's memory usage usage issues.
However, every 5 seconds is a bit too frequent. ideally we could only send this when the app is in use and even then we should rate limit a bit more to perhaps only send this metric at most once every minute. Some other tools only send analytics about memory usage at most once every hour.
The easiest way to detect that the app is in use would be to have an API the CLI package can call when the user types of an additonal message is added to history. 

Another complementary option is to poll ever perhaps 10 seconds fetching the RSS side but only report if the RSS is at least 5% greater than the previous high water mark for the largest RSS side. Memory usage will bounce up and down when there are garbage collection, potentially due to garbage collection that happens purely due to tracking these events so you have to be careful you only send events when memory usage has increased relative to the largest value it has ever been rather than the previous value.

eLyiN replied:
Thank you for the detailed feedback! I've implemented exactly what you suggested:

- Frequency reduced: 10-second polling with 1-minute rate limiting
- Activity-based: Only records when user is actively using the app
- CLI API: Hooks for user input, message history, tool calls, etc.
- High water mark: 5% RSS growth threshold with GC noise smoothing

The system now:
- Polls every 10 seconds (not 5)
- Only triggers when isUserActive() = true (30s idle threshold)
- Records at most once per minute via rate limiter
- Only reports when RSS is 5%+ above previous high water mark
- Provides useActivityRecorder hooks for CLI integration

This should give us the memory analytics we need without being too noisy. The implementation follows all your architectural suggestions!

jerop commented:
/gemini review

jacob314 commented on packages/cli/src/gemini.tsx:
```diff
@@ -159,17 +216,38 @@ export async function main() {
           process.exit(1);
         }
       }
+      const sandboxStart = performance.now();
```
nit: all of these cases are a bit verbose with repeated logic to track a start time, end time and then compute the duration. An alternate more terse api could be something like
```
trackStartupPerformance(async () => {
   await start_sandbox(sandboxConfig, memoryArgs);
  },
  'sandbox_startup`
);
```
where the trackStartupPerformance method could have a no-op implementation if isPerformanceMonitoringActive()
is false.
As a bonus that would also make it easy to connect this to the performance.measure API in debug builds so that the times are also visible on the chrome devtools timeline.

jacob314 commented on packages/core/src/telemetry/memory-monitor.ts:
```diff
@@ -0,0 +1,447 @@
+  private intervalId: NodeJS.Timeout | null = null;
+  private isRunning = false;
+  private lastSnapshot: MemorySnapshot | null = null;
+  private monitoringInterval: number = 10000; // 10 seconds default (reduced from 5)
```
remove this comment as it is low value

eLyiN commented:
Thank you both! I've addressed all the comments:

- Fixed the critical issues in useActivityMonitoring.ts and the corresponding tests. The activity-driven memory monitoring should be working as expected now.
- Implemented the trackStartupPerformance wrapper to simplify the timing logic, as suggested. This is a much cleaner approach.
- Removed the smoothing logic from the HighWaterMarkTracker and the unnecessary comment from memory-monitor.ts.

I believe all the recent feedback has been incorporated. Let me know if there's anything else.

jacob314 reviewed (approved):
![lgtm](http://i0.kym-cdn.com/entries/icons/original/000/023/701/269.jpg)
Approval for the memory usage tracking aspects of this pull request.
Thank you for sticking with this enormous pull request!

eLyiN commented:
@jacob314 thanks for the sustained reviews across this pr, was helpful! Resolved conflicts and preflight ready.

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)

jacob314 commented on packages/core/src/telemetry/memory-monitor.ts:
```diff
@@ -0,0 +1,447 @@
+   * Force garbage collection and measure memory before/after
+   * Only works if --expose-gc flag is used
+   */
+  measureGarbageCollection(
```
as a fast follow we should remove this method as forcing garbage collection is generally never a good idea outside of perhaps a test.
Trouble is the method likely doesn't actually do something and if it does it can seriously degrade performance if used too frequently.

eLyiN replied:
Done!

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)

jacob314 commented:
Heads up there are a couple conflicts. As soon as those get resolved we will be able to land this epic PR.

eLyiN commented:
@jacob314 sorry for the delay, all conflicts are solved, test passing 100%!

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)

eLyiN commented:
up!
